### PR TITLE
refactor(decimal): allow the user to input any value - FE-3705

### LINKED ISSUE #3372: Decimal fields crashes app if "^" is used on Mac.
### Current behaviour
If you try to type ^ into a decimal carbon field on a Mac (Firefox, Chrome) it will crash the react app with an error saying that your entry is not a decimal number.

### Expected behaviour
Decimal fields should also permit typing of ^ on Mac, as it does on windows.

### Your environment
carbon-react Version 48.6.2
macOS Catalina 10.15.6
Firefox 82.0.3

### PR DESCRIPTION
**Pull request by DipperTheDan**
BREAKING CHANGE: 
Precision prop can no longer be changed on the fly. Your app will need to refresh in order for the user to see the new precision value.

### Proposed behaviour

- Allow user to input anything they want in to Decimal component
- Remove styled wiggle

### Current behaviour

- Special and Alphabetic characters are currently prevented from being input
- Component currently has styled wiggle to inform the user they have tried to input an illegal character

### Checklist

- [x] Commits follow our style guide
<del> - [ ] Screenshots are included in the PR if useful </del>
- [x] All themes are supported if required
- [x] Unit tests added or updated if required
- [x] Cypress automation tests added or updated if required
- [x] Storybook added or updated if required
- [x] Typescript `d.ts` file added or updated if required
- [x] Carbon implementation and Design System documentation are congruent

### Additional context
fixes #3372 

### Testing instructions
- Manual testing - https://codesandbox.io/s/carbon-quickstart-forked-01vep

- Locale testing - https://codesandbox.io/s/carbon-quickstart-forked-1v8qj

The below table is to give an idea of what manual tests should be used to check the Decimal component. 

| Precision | User Input | Value  |Status|
|-----------|------------|--------|-----|
|(Default)     | @£$%^         | @£$%^      |
|(Default)    | ^          | ^      |
|(Default)      | ˇ         | ˇ    |
|(Default)    | ---       | ---    |
| 0         | 1          | 1      |
| 0         | 2          | 2      |
| 0         | 1234567789 | 1,234,567,789 |
| 1         | 1          | 1.0    |
| 1         | 1.2        | 1.2    |
| 1         | 1.23       | 1.23   |
| 2         | 2          | 2.00   |
| 2         | 2.1        | 2.10   |
| 2         | 2.12       | 2.12   |
| 2         | 2.123      | 2.123  |
| 3         | 2.1        | 2.100  |
| 3         | 2.12       | 2.120  |
| 3         | 2.123      | 2.123  |
| 3         | 2.1234     | 2.1234 |
| 4         | 1.1234         | 1 .1234     |
| 4         | 2          | 2.0000      |
| 4         | 234556654          | 234,556,654.0000      |
| 5         | 1          | 1.00000    |
| 5         | 1.12345        | 1.12345    |
| 5         | 1a.23       | 1a.23   |
| 6         | 2.123456          | 2.123456   |
| 6         | 2.1        | 2.1000000   |
| 6         | 2a.12       | 2a.12   |
| 6         | 1,232.123      | 1,232.123000  |
| 7         | 1          | 1.0000000      |
| 7         | 2344.1234567  | 2,344.1234567 |
| 7         | 88652344.1234567  | 88,652,344.1234567 |
| 8         | 1          | 1.00000000    |
| 8         | 1.2        | 1.20000000    |
| 8         | 1.23       | 1.23000000   |
| 9         | 2          | 2.000000000   |
| 9         | 2.1        | 2.100000000   |
| 9         | 1222.12       | 1,222.120000000   |
| 9         | 2.123000000      | 2.123000000  |
| 10         | 2.1        | 2.1000000000  |
| 10         | 2.12       | 2.1200000000  |
| 10         | 1222.123      | 1,222.1230000000  |
| 10         | 2.12345     | 2.1234500000 |
| 11         | 1         | 1.00000000000    |
| 11         | 2345          | 2,345.00000000000      |
| 11         | 17899536472345          | 17,899,536,472,345.00000000000      |
| 12         | 1          | 1.000000000000    |
| 12         | 1.12345        | 1.123450000000    |
| 12         | 1a.23       | 1a.23   |
| 13         | 2          | 2.0000000000000   |
| 13         | 2.1        | 2.1000000000000   |
| 13         | 2a.12       | 2a.12   |
| 13         | 1,232.123      | 1,232.1230000000000  |
| 14         | 2.1        | 2.10000000000000  |
| 14         | 2.12       | 2.12000000000000  |
| 14         | 2.123      | 2.12300000000000  |
| 14         | 1222.1234     | 1,222.12340000000000 |
| 15         | 1         | 1.000000000000000  |
| 15         | 2332.78          | 2,332.780000000000000      |
| 15         | 1a3.55          | 1a3.55    |
| 15         | 1.12345        | 1.123450000000000    |
| 15         | 1a.23       | 1a.23   |
| 16         | 2.123456          | 2.1234560000000000 && Prop validation error to appear in console|

### CODE REVIEW

File: src/__experimental__/components/decimal/decimal.component.js
Comment (edleeks87) [inline src/__experimental__/components/decimal/decimal.component.js:156]: Do we want to remove these checks? It might be worth moving it into a custom prop type function

File: src/__experimental__/components/decimal/decimal.spec.js
Comment (chrisbarber86) [inline src/__experimental__/components/decimal/decimal.spec.js:744]: Should we update hiddenValue to be FooBar? Normally the expectation of the hiddenValue is that it is strictly a plainly formatted (i.e. only the seperator & numbers present) value. If the typed value is `FooBar` i'd expect the hiddenValue to be `null`

File: src/__experimental__/components/decimal/decimal.component.js
Comment (chrisbarber86) [inline src/__experimental__/components/decimal/decimal.component.js:83]: Won't the logic in here need to update as well?

**Review by chrisbarber86 — CHANGES_REQUESTED** [review]

File: src/__experimental__/components/decimal/decimal.spec.js
Comment (DipperTheDan) [reply src/__experimental__/components/decimal/decimal.spec.js:744]: My thoughts on this were that the consumer would be using the hiddenValue for validation purposes. I'm happy to change it to `null` to meet the expectation of it being a plainly formatted value though.

File: src/__experimental__/components/decimal/decimal.component.js
Comment (DipperTheDan) [reply src/__experimental__/components/decimal/decimal.component.js:8]: I agree. I'll make sure we have a ticket on our backlog for it 😄

File: src/__experimental__/components/decimal/decimal.component.js
Comment (edleeks87) [inline src/__experimental__/components/decimal/decimal.component.js:227]: I think you need to check if it's a integer number as well here

```suggestion
    if (!Number.isInteger(precision) || props.precision < 0 || props.precision > Decimal.maxPrecision) {
```

**Review by edleeks87 — DISMISSED** [review]

File: src/__experimental__/components/decimal/decimal.spec.js
Comment (jamime) [inline src/__experimental__/components/decimal/decimal.spec.js:598]: Instead of deleting these tests I think we can assert on the value instead, to make sure the value changed correctly. Same for the tests below.

**Review by jamime — CHANGES_REQUESTED** [review]
There is an inconsistency between how the user input `2` and `3.00` is handled, see the video below. I think these should be consistent.


https://user-images.githubusercontent.com/2328042/111334141-37292c80-866b-11eb-9d93-b639751fa2a8.mov

I added this to the render method to debug

```diff
  render() {
    const { name, defaultValue, ...rest } = this.props;
    return (
      <>
+      <pre>{JSON.stringify(this.state, null, 2)}</pre>
        <Textbox
```

**Comment by DipperTheDan** [comment]
> There is an inconsistency between how the user input `2` and `3.00` is handled, see the video below. I think these should be consistent.
> 

This is how Decimal behaves in `master` so I left it that way as I was unsure if changing this would have an effect on existing implementations of Decimal. I've attached a video of this in `master`.

Uploading Decimal Value Recording.mov…

**Comment by jamime** [comment]
> > There is an inconsistency between how the user input `2` and `3.00` is handled, see the video below. I think these should be consistent.
> 
> @jamime This is how Decimal behaves in `master` so I left it that way as I was unsure if changing this would have an effect on existing implementations of Decimal. I've attached a video of this in `master`.
> 
>  Decimal.Value.Recording.mov

Ah ok, we can address that at a later date.

**Review by jamime — CHANGES_REQUESTED** [review]
There are some requirements that have not yet been fulfilled.

> Field should still auto format on blur adding the the internationalised separators, delimiters and trailing .00, 00, 0. This is configurable using the precision prop.

The PR still applies **rounding** to the precision limit. We don't want to round at all. We do want to append `0`'s where the value was less than the precision limit. These are the same value, but with a different representation `1.2 === 1.20 === 1.200`.

> Field should allow any character to be inputted, the consumer should then provide their own decimal validation.

When there is more decimal places than the precision limit we should not modify the value.

Here are some examples that illustrate this

| Precision | User Input | Value  |
|-----------|------------|--------|
| 0         | 1          | 1      |
| 0         | 2          | 2      |
| 1         | 1          | 1.0    |
| 1         | 1.2        | 1.2    |
| 1         | 1.23       | 1.23   |
| 2         | 2          | 2.00   |
| 2         | 2.1        | 2.10   |
| 2         | 2.12       | 2.12   |
| 2         | 2.123      | 2.123  |
| 3         | 2.1        | 2.100  |
| 3         | 2.12       | 2.120  |
| 3         | 2.123      | 2.123  |
| 3         | 2.1234     | 2.1234 |

Please let me know if you have any questions.

File: src/__experimental__/components/decimal/decimal.spec.js
Comment (jamime) [reply src/__experimental__/components/decimal/decimal.spec.js]: I've just reviewed this @chrisbarber86 but in this case I think we should actually send `FooBar`. If we were to send a `POST` with nothing as the value the backend could only say `this field is required`, but not `FooBar isn't valid`.

If the field was optional, and we sent nothing the backend wouldnt be able to flag that the input value was discarded, which is the same issue we're trying to avoid in SPA validations.

File: src/__experimental__/components/decimal/decimal.component.js
Comment (jamime) [reply src/__experimental__/components/decimal/decimal.component.js:83]: I think we can remove all of the `onBlur` code because it is only handling the `-` key press. If a user pressed `-` accidentally instead of `9` we could pick up the user error in validations because we're reverting to the `defaultValue` currently.

File: src/__experimental__/components/decimal/decimal.spec.js
Comment (chrisbarber86) [reply src/__experimental__/components/decimal/decimal.spec.js:940]: Ok - so the consumer on the front end would need to add validations if they don't want to send `FooBar`. otherwise if they don't it's on the server to handle that.

I'm ok with that.

File: src/__experimental__/components/decimal/decimal.component.js
Comment (jamime) [inline src/__experimental__/components/decimal/decimal.component.js:170]: If you enter `1234.56` it will be re-formatted as `1,234` and will loose precision. We don't want this to happen. Please add a test that covers this.

**Review by jamime — CHANGES_REQUESTED** [review]

File: cypress/features/regression/experimental/decimalPrecision.feature
Comment (DipperTheDan) [reply cypress/features/regression/experimental/decimalPrecision.feature:12]: Removed by mistake. Added back in 👍🏻

File: src/__experimental__/components/decimal/decimal.spec.js
Comment (DipperTheDan) [reply src/__experimental__/components/decimal/decimal.spec.js:272]: Test still exists but has been rewritten as this is no longer an invariant but a console error. 
https://github.com/Sage/carbon/pull/3732/files/e815557daffb13aaea8b4d92c03e5b341b0f72b3#diff-992c7dd22ced4f0d4c1a51cd13658c283bbd2e5d4e56e776892de6672be2e369R215

File: src/__experimental__/components/decimal/decimal.component.js
Comment (DipperTheDan) [reply src/__experimental__/components/decimal/decimal.component.js:170]: Test has been added to cover this scenario.

File: src/__experimental__/components/decimal/decimal.spec.js
Comment (DipperTheDan) [reply src/__experimental__/components/decimal/decimal.spec.js:319]: Test has been added. Good shout thinking about adding this 👍🏻

File: src/__experimental__/components/decimal/decimal.spec.js
Comment (DipperTheDan) [reply src/__experimental__/components/decimal/decimal.spec.js:225]: Messaged you privately about this as it required a screen recording and a lengthy message that could trigger a discussion.

File: src/__experimental__/components/decimal/decimal.component.js
Comment (jamime) [inline src/__experimental__/components/decimal/decimal.component.js:37]: This isn't true any more. 

```suggestion
        Changing the "Decimal `precision` prop has no effect.
```

File: cypress/features/regression/experimental/decimalPrecision.feature
Comment (jamime) [inline cypress/features/regression/experimental/decimalPrecision.feature:29]: Copy manual tests into here

File: src/__experimental__/components/decimal/decimal.component.js
Comment (jamime) [inline src/__experimental__/components/decimal/decimal.component.js:160]: I've sent some code on slack to update this with.

**Review by jamime — CHANGES_REQUESTED** [review]
After these changes I'll re-review and approve and we'll then squash the commits and add a `BREAKING CHANGE` for the precision prop (and anything else).

File: src/__experimental__/components/decimal/decimal.component.js
Comment (DipperTheDan) [reply src/__experimental__/components/decimal/decimal.component.js:160]: 👍🏻 Applied.

File: src/__experimental__/components/decimal/decimal.component.js
Comment (DipperTheDan) [reply src/__experimental__/components/decimal/decimal.component.js:37]: Changed this error message.

**Review by jamime — APPROVED** [review]

File: cypress/features/regression/experimental/decimalPrecision.feature
Comment (DlgSHi) [inline cypress/features/regression/experimental/decimalPrecision.feature:12]: This TC is the same as a previous one, we can change it into some character instead of checking the same number row.

Comment (DipperTheDan) [reply cypress/features/regression/experimental/decimalPrecision.feature:12]: That is a good idea. I'll use this replacement as a `dead key` test. Cheers for spotting that 👍🏻

**Review by edleeks87 — APPROVED** [review]

**Review by jamime — APPROVED** [review]

**Review by chrisbarber86 — APPROVED** [review]

**Review by edleeks87 — APPROVED** [review]

**Comment by harpalsingh** [comment]
As per the requirements, I have checked all expected outcomes and accessibility requirements to enter any character are working as expected/

File: src/__experimental__/components/decimal/decimal.component.js
Comment (jamime) [inline src/__experimental__/components/decimal/decimal.component.js:146]: The delimiter isn't always `,`, it depends on the users locale.

**Review by jamime — CHANGES_REQUESTED** [review]

File: src/__experimental__/components/decimal/decimal.component.js
Comment (edleeks87) [reply src/__experimental__/components/decimal/decimal.component.js:146]: This will match any special (non-alphanumeric) characters so I think just the comment needs updating?

**Review by jamime — COMMENTED** [review]
@DipperTheDan is currently working on defects raised by QA. When using the FR locale the remainder is rounded if it is greater than the precision.

**Review by jamime — COMMENTED** [review]
We reviewed the changes to date and made some minor ammendments to the I18NHelpers. You can now use `I18nHelper.formatDecimal(12345.678, 2, {round: false})` which will format the decimal as `12,345.687`. The standard implementation would return `12,345.69`. This has allowed us to remove the duplicated code in `Decimal.component.js`.

There is one outstanding issue that @DipperTheDan is going to look into.

- When the locale is `fr`
- and the user enters `123,,56` (123.56)
- value is changed to `123..56` (12,356)

These are both invalid decimals, but it is preferred if the value does not automatically change.

File: cypress/support/step-definitions/decimal-steps.js
Comment (jamime) [inline cypress/support/step-definitions/decimal-steps.js:10]: Why do we need to wait?

**Review by jamime — COMMENTED** [review]
👍  LGTM. I'll approve once commits have been squashed (after QA).

**Comment by edleeks87** [comment]
👍 LGTM. I'll approve once commits have been squashed (after QA).

**Review by DlgSHi — APPROVED** [review]

**Review by jamime — APPROVED** [review]

**Review by edleeks87 — APPROVED** [review]

**Pull request merged**
